File: experiments-local/experiment-mu-glm52-ompfix/runs/agent=omp_language=typescript_model=openrouter/z-ai/glm-5.2_tooling=none/rep3/src/tools.ts
/**
 * Brazilian Soccer MCP Server — Tool Definitions
 * -----------------------------------------------------------------------------
 * Context block:
 *   This module registers the MCP tools exposed by the server. Each tool maps
 *   to one of the five required capability categories in the spec:
 *     1. search_matches      — Match Queries (by team, opponent, date range,
 *        competition, season, stage)
 *     2. team_stats           — Team Queries (W/D/L, goals, win rate, venue)
 *     3. head_to_head         — Team Queries (head-to-head comparison)
 *     4. standings            — Competition Queries (computed table + champion)
 *     5. match_statistics     — Statistical Analysis (avg goals, home/away
 *        win rates, biggest wins)
 *     6. biggest_wins         — Statistical Analysis (ranked victories)
 *     7. last_match           — Simple Lookups ("when did X last play Y?")
 *     8. search_players       — Player Queries (name/nationality/club/position)
 *     9. top_players          — Player Queries (ranked by overall)
 *    10. brazilian_players_at_brazilian_clubs — Player Queries (cross-file
 *        Brazilian-clubs grouping)
 *    11. list_competitions    — Discovery (catalog of loaded datasets)
 *
 *   Every tool returns a single `content` text blob produced by the pure
 *   formatters in `format.ts`. The query engine does the work; tools only
 *   translate args ↔ filters and format. Schemas use zod raw shapes passed
 *   to `registerTool`. All tools are read-only.
 */

import { z } from "zod";
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import type { Dataset } from "./data/types.js";
import {
  biggestWins,
  brazilianPlayersAtBrazilianClubs,
  findMatches,
  findPlayers,
  headToHead,
  lastMatch,
  matchStatistics,
  standings,
  teamStats,
} from "./data/query.js";
import {
  formatBiggestWins,
  formatClubBrazilianPlayers,
  formatHeadToHead,
  formatMatch,
  formatMatches,
  formatPlayers,
  formatStandings,
  formatStatistics,
  formatTeamStats,
} from "./data/format.js";

/** Competition keys accepted by filtering tools. */
const COMPETITION_VALUES = [
  "brasileirao",
  "copa-do-brasil",
  "libertadores",
  "brasileirao-historical",
  "serie-a",
  "serie-b",
  "serie-c",
  "copa-do-brasil-ext",
] as const;

/** Register all Brazilian-soccer tools on the given MCP server. */
export function registerTools(server: McpServer, ds: Dataset): void {
  // ----- 1. search_matches -----
  server.registerTool(
    "search_matches",
    {
      title: "Search matches",
      description:
        "Find football matches by team, opponent, venue, competition, season, date range, or stage. Returns a formatted list (newest first).",
      inputSchema: {
        team: z.string().optional().describe("Team name (tolerant: 'Flamengo' matches 'Flamengo-RJ')."),
        opponent: z.string().optional().describe("Opponent team name; combined with team for head-to-head-style filtering."),
        venue: z.enum(["home", "away", "either"]).optional().describe("Venue filter for 'team'."),
        competition: z.enum(COMPETITION_VALUES).optional().describe("Competition to scope to."),
        season: z.number().int().optional().describe("Season year."),
        from: z.string().optional().describe("ISO date YYYY-MM-DD, inclusive lower bound."),
        to: z.string().optional().describe("ISO date YYYY-MM-DD, inclusive upper bound."),
        stage: z.string().optional().describe("Tournament stage, e.g. 'final', 'group stage' (Libertadores)."),
        round: z.string().optional().describe("Round number/label."),
        limit: z.number().int().min(0).optional().describe("Max matches to return (default 50)."),
      },
    },
    (args) => {
      const matches = findMatches(ds, {
        team: args.team,
        opponent: args.opponent,
        venue: args.venue,
        competition: args.competition,
        season: args.season,
        from: args.from,
        to: args.to,
        stage: args.stage,
        round: args.round,
        limit: args.limit ?? 50,
      });
      return text(formatMatches(matches, "Matches"));
    },
  );

  // ----- 2. team_stats -----
  server.registerTool(
    "team_stats",
    {
      title: "Team statistics",
      description:
        "Compute win/draw/loss record, goals scored/conceded, points and win rate for a team, optionally scoped by competition, season, venue and date range.",
      inputSchema: {
        team: z.string().describe("Team name (tolerant normalization applied)."),
        competition: z.enum(COMPETITION_VALUES).optional(),
        season: z.number().int().optional(),
        venue: z.enum(["home", "away", "either"]).optional(),
        from: z.string().optional(),
        to: z.string().optional(),
      },
    },
    (args) => {
      const stats = teamStats(ds, args.team, {
        competition: args.competition,
        season: args.season,
        venue: args.venue,
        from: args.from,
        to: args.to,
      });
      const scope = scopeLabel(args);
      return text(formatTeamStats(stats.team, stats, scope));
    },
  );

  // ----- 3. head_to_head -----
  server.registerTool(
    "head_to_head",
    {
      title: "Head-to-head",
      description:
        "Compare two teams head-to-head across the dataset: wins, draws, goals, and a sample of matches. Handles team-name variations and either venue/order.",
      inputSchema: {
        team1: z.string().describe("First team name."),
        team2: z.string().describe("Second team name."),
        competition: z.enum(COMPETITION_VALUES).optional(),
        season: z.number().int().optional(),
      },
    },
    (args) => {
      const h2h = headToHead(ds, args.team1, args.team2, {
        competition: args.competition,
        season: args.season,
      });
      return text(formatHeadToHead(h2h));
    },
  );

  // ----- 4. standings -----
  server.registerTool(
    "standings",
    {
      title: "Competition standings",
      description:
        "Compute a league standings table (points, W/D/L, goals, GD) from match results, scoped by competition and season. Returns ranked table with the champion.",
      inputSchema: {
        competition: z.enum(COMPETITION_VALUES).describe("Competition to compute standings for."),
        season: z.number().int().describe("Season year."),
      },
    },
    (args) => {
      const rows = standings(ds, { competition: args.competition, season: args.season });
      const title = `Standings — ${args.competition} ${args.season}`;
      return text(formatStandings(rows, title));
    },
  );

  // ----- 5. match_statistics -----
  server.registerTool(
    "match_statistics",
    {
      title: "Match statistics",
      description:
        "Aggregate statistics over a scoped set of matches: average goals, home/away win & draw rates, and biggest home/away wins. Scope by competition and/or season.",
      inputSchema: {
        competition: z.enum(COMPETITION_VALUES).optional(),
        season: z.number().int().optional(),
        team: z.string().optional().describe("Restrict to matches involving this team."),
      },
    },
    (args) => {
      const stats = matchStatistics(ds, {
        competition: args.competition,
        season: args.season,
        team: args.team,
      });
      return text(formatStatistics(stats, scopeLabel(args)));
    },
  );

  // ----- 6. biggest_wins -----
  server.registerTool(
    "biggest_wins",
    {
      title: "Biggest victories",
      description: "List the biggest victory margins in a scoped set of matches (competition/season/team).",
      inputSchema: {
        competition: z.enum(COMPETITION_VALUES).optional(),
        season: z.number().int().optional(),
        team: z.string().optional(),
        limit: z.number().int().min(1).max(100).optional().describe("Number of results (default 10)."),
      },
    },
    (args) => {
      const wins = biggestWins(ds, {
        competition: args.competition,
        season: args.season,
        team: args.team,
      }, args.limit ?? 10);
      return text(formatBiggestWins(wins, "Biggest victories"));
    },
  );

  // ----- 7. last_match -----
  server.registerTool(
    "last_match",
    {
      title: "Last match",
      description: "Find the most recent match involving a team (optionally vs an opponent, scoped by competition/season).",
      inputSchema: {
        team: z.string().describe("Team name."),
        opponent: z.string().optional(),
        competition: z.enum(COMPETITION_VALUES).optional(),
        season: z.number().int().optional(),
      },
    },
    (args) => {
      const m = lastMatch(ds, args.team, {
        opponent: args.opponent,
        competition: args.competition,
        season: args.season,
      });
      return m ? text(formatMatch(m, "Last match")) : text("No match found.");
    },
  );

  // ----- 8. search_players -----
  server.registerTool(
    "search_players",
    {
      title: "Search players",
      description:
        "Search the FIFA player database by name, nationality, club, position and/or minimum overall rating. Results are ranked by overall rating.",
      inputSchema: {
        name: z.string().optional(),
        nationality: z.string().optional().describe("e.g. 'Brazil'."),
        club: z.string().optional(),
        position: z.string().optional().describe("e.g. 'ST', 'GK', 'LW'."),
        minOverall: z.number().int().optional(),
        limit: z.number().int().min(1).optional().describe("Max results (default 20)."),
      },
    },
    (args) => {
      const players = findPlayers(ds, {
        name: args.name,
        nationality: args.nationality,
        club: args.club,
        position: args.position,
        minOverall: args.minOverall,
        limit: args.limit ?? 20,
      });
      return text(formatPlayers(players, "Players"));
    },
  );

  // ----- 9. top_players -----
  server.registerTool(
    "top_players",
    {
      title: "Top players",
      description: "List the highest-rated players, optionally filtered by nationality (e.g. Brazilian players) or club.",
      inputSchema: {
        nationality: z.string().optional(),
        club: z.string().optional(),
        limit: z.number().int().min(1).optional().describe("Number of results (default 10)."),
      },
    },
    (args) => {
      const players = findPlayers(ds, {
        nationality: args.nationality,
        club: args.club,
        limit: args.limit ?? 10,
      });
      const title = args.nationality
        ? `Top ${args.nationality} players`
        : args.club
          ? `Top players at ${args.club}`
          : "Top players";
      return text(formatPlayers(players, title));
    },
  );

  // ----- 10. brazilian_players_at_brazilian_clubs -----
  server.registerTool(
    "brazilian_players_at_brazilian_clubs",
    {
      title: "Brazilian players at Brazilian clubs",
      description: "Group Brazilian players from the FIFA dataset by Brazilian club, with counts and average ratings.",
      inputSchema: {},
    },
    () => text(formatClubBrazilianPlayers(brazilianPlayersAtBrazilianClubs(ds))),
  );

  // ----- 11. list_competitions -----
  server.registerTool(
    "list_competitions",
    {
      title: "List competitions",
      description: "List all loaded datasets/competitions with their seasons and match counts.",
      inputSchema: {},
    },
    () => {
      const lines = ["Loaded competitions:"];
      for (const c of ds.competitions) {
        const seasons = c.seasons.length ? c.seasons.join(", ") : "n/a";
        lines.push(`- ${c.label} [${c.competition}] (${c.source}): ${c.matchCount} matches, seasons: ${seasons}`);
      }
      lines.push(`- Players: ${ds.players.length}`);
      return text(lines.join("\n"));
    },
  );
}

/** Build a short human scope label for a stats call. */
function scopeLabel(args: { competition?: string; season?: number; venue?: string }): string {
  const parts: string[] = [];
  if (args.competition) parts.push(args.competition);
  if (args.season != null) parts.push(String(args.season));
  if (args.venue) parts.push(args.venue);
  return parts.join(" ");
}

/** Wrap a string as a text CallToolResult. */
function text(content: string) {
  return { content: [{ type: "text" as const, text: content }] };
}

File: experiments-local/experiment-mu-glm53-provider/runs/agent=oc-fireworks_language=python_model=openrouter/z-ai/glm-5.3_tooling=none/rep1/brazilian_soccer_mcp/tools.py
"""
 brazilian_soccer_mcp / tools.py
 ===============================

 Why
 ---
 The MCP server and the tests need one authoritative catalogue of the
 server's tools: names, human descriptions, JSON-Schema input schemas and
 the dispatch from a tool call to a query result plus rendered text.
 Defining it here keeps server.py a pure protocol layer.

 What
 ---
 * :data:`TOOLS`            - tool descriptors for MCP ``tools/list``.
 * :func:`call_tool`        - dispatch (name, arguments) -> result dict
                              (``{"content", "structuredContent",
                              "isError"}``), raising :class:`ToolError`
                              for unknown tools / bad arguments.
 * :func:`tool_summaries`   - one-line help text for README/CLI.

 Tools (one per TASK.md capability):
   search_matches, last_match_between, head_to_head, team_stats,
   team_profile, standings, biggest_wins, competition_stats,
   best_records, derbies, player_search, player_club_report,
   list_competitions, list_teams

 Test: ``tests/test_server.py`` (descriptor shapes + dispatch).
=====================================================================
"""

from __future__ import annotations

from . import queries, render
from .loader import Dataset


class ToolError(Exception):
    """Raised for unknown tools or malformed arguments (protocol error)."""


def _schema(properties: dict, required: list[str] | None = None) -> dict:
    return {
        "type": "object",
        "properties": properties,
        "required": required or [],
        "additionalProperties": False,
    }


_STR = {"type": "string"}
_INT = {"type": "integer", "minimum": 1}


TOOLS: list[dict] = [
    {
        "name": "search_matches",
        "description": (
            "Search Brazilian soccer fixtures by team (any side), opponent, "
            "competition (serie_a, serie_b, serie_c, copa_do_brasil, "
            "libertadores, or 'all'), season, stage/round (e.g. 'final', "
            "'semifinals', 'group stage', or a numeric round) and date range "
            "(ISO YYYY-MM-DD). Returns newest-first matches with date, score, "
            "competition and phase. Handles every team-name spelling used by "
            "the datasets (e.g. 'Palmeiras-SP' = 'Palmeiras')."
        ),
        "inputSchema": _schema(
            {
                "team": {"type": "string", "description": "Team name (any spelling)"},
                "opponent": {
                    "type": "string",
                    "description": "Restrict to fixtures vs this team",
                },
                "competition": {
                    "type": "string",
                    "description": "Competition id or alias",
                },
                "season": {"type": "integer", "description": "Year, e.g. 2019"},
                "stage": {
                    "type": "string",
                    "description": "'final', 'semifinals', 'group stage' or round number",
                },
                "date_from": {"type": "string", "description": "ISO date, inclusive"},
                "date_to": {"type": "string", "description": "ISO date, inclusive"},
                "limit": {
                    **_INT,
                    "maximum": 200,
                    "description": "Max matches to return (default 20)",
                },
            }
        ),
    },
    {
        "name": "last_match_between",
        "description": (
            "Most recent played fixture between two teams (date, score, "
            "competition), plus any later scheduled-but-unplayed fixture."
        ),
        "inputSchema": _schema(
            {
                "team_a": {**_STR, "description": "First team"},
                "team_b": {**_STR, "description": "Second team"},
            },
            required=["team_a", "team_b"],
        ),
    },
    {
        "name": "head_to_head",
        "description": (
            "Head-to-head record between two teams: meetings, wins/draws/losses, "
            "goals, and the fixture list (optionally scoped to a competition "
            "and season)."
        ),
        "inputSchema": _schema(
            {
                "team_a": {**_STR, "description": "First team"},
                "team_b": {**_STR, "description": "Second team"},
                "competition": _STR,
                "season": {"type": "integer"},
                "limit": _INT,
            },
            required=["team_a", "team_b"],
        ),
    },
    {
        "name": "team_stats",
        "description": (
            "Win/draw/loss record and goals for one team, scoped by season, "
            "competition and venue ('home', 'away', 'all'). Includes home/away "
            "splits and per-competition breakdowns."
        ),
        "inputSchema": _schema(
            {
                "team": {**_STR, "description": "Team name (any spelling)"},
                "season": {"type": "integer"},
                "competition": _STR,
                "venue": {"type": "string", "enum": ["home", "away", "all"]},
            },
            required=["team"],
        ),
    },
    {
        "name": "team_profile",
        "description": (
            "Everything known about one club: canonical id, state, every name "
            "spelling seen in the data, competitions and seasons played, "
            "all-time record, and FIFA-database player presence. Answers "
            "'what competitions has Palmeiras played in?'."
        ),
        "inputSchema": _schema(
            {
                "team": {**_STR, "description": "Team name (any spelling)"},
            },
            required=["team"],
        ),
    },
    {
        "name": "standings",
        "description": (
            "League table computed from match results (3 pts/win, CBF "
            "tie-breaks) with champion and relegated (bottom four) teams. "
            "League competitions only - cups have no table."
        ),
        "inputSchema": _schema(
            {
                "competition": {
                    **_STR,
                    "description": "serie_a, serie_b or serie_c (or alias)",
                },
                "season": {"type": "integer", "description": "Year, e.g. 2019"},
            },
            required=["competition", "season"],
        ),
    },
    {
        "name": "biggest_wins",
        "description": (
            "Largest victory margins in the dataset (all competitions or one, "
            "optionally one season), ranked by goal margin then total goals."
        ),
        "inputSchema": _schema(
            {
                "competition": _STR,
                "season": {"type": "integer"},
                "limit": _INT,
            }
        ),
    },
    {
        "name": "competition_stats",
        "description": (
            "Aggregate statistics: matches, average goals per match, home/draw/"
            "away win rates - for one competition (+season) or overall with a "
            "per-competition breakdown."
        ),
        "inputSchema": _schema(
            {
                "competition": _STR,
                "season": {"type": "integer"},
            }
        ),
    },
    {
        "name": "best_records",
        "description": (
            "Rank teams by performance for a venue: win_rate, points_per_game, "
            "goals_for, goals_against or avg_goals_for (home/away/all). "
            "Answers 'which team has the best away record?'."
        ),
        "inputSchema": _schema(
            {
                "venue": {"type": "string", "enum": ["home", "away", "all"]},
                "competition": _STR,
                "season": {"type": "integer"},
                "metric": {
                    "type": "string",
                    "enum": [
                        "win_rate",
                        "points_per_game",
                        "goals_for",
                        "goals_against",
                        "avg_goals_for",
                    ],
                },
                "min_matches": {"type": "integer", "minimum": 1},
                "limit": _INT,
            }
        ),
    },
    {
        "name": "derbies",
        "description": (
            "Classic Brazilian derbies (Fla-Flu, Gre-Nal, Derby Paulista, "
            "Ba-Vi, ...) with all-time head-to-head records and fixtures, "
            "optionally for one season. Answers 'show me all derbies in 2023'."
        ),
        "inputSchema": _schema(
            {
                "season": {"type": "integer"},
                "limit_per_derby": _INT,
            }
        ),
    },
    {
        "name": "player_search",
        "description": (
            "Search the FIFA player database (18,207 players): name substring, "
            "nationality (e.g. Brazil), club, position (code like ST/LW/GK or "
            "group DEF/MID/FWD), minimum overall rating, maximum age. Ordered "
            "by overall/potential/age/name/value."
        ),
        "inputSchema": _schema(
            {
                "name": _STR,
                "nationality": _STR,
                "club": _STR,
                "position": _STR,
                "min_overall": {"type": "integer", "minimum": 0, "maximum": 99},
                "max_age": {"type": "integer", "minimum": 15, "maximum": 60},
                "order": {
                    "type": "string",
                    "enum": ["overall", "potential", "age", "name", "value"],
                },
                "limit": _INT,
            }
        ),
    },
    {
        "name": "player_club_report",
        "description": (
            "Players grouped by club: counts, average and best ratings, "
            "flagging clubs the match data knows as Brazilian. Answers "
            "'Brazilian players at Brazilian clubs'."
        ),
        "inputSchema": _schema(
            {
                "nationality": _STR,
                "min_overall": {"type": "integer", "minimum": 0, "maximum": 99},
            }
        ),
    },
    {
        "name": "list_competitions",
        "description": "List every competition with seasons covered and match counts.",
        "inputSchema": _schema({}),
    },
    {
        "name": "list_teams",
        "description": (
            "Team directory: all clubs, or participants of one competition "
            "and season, with match counts and canonical ids."
        ),
        "inputSchema": _schema(
            {
                "competition": _STR,
                "season": {"type": "integer"},
                "limit": _INT,
            }
        ),
    },
]

TOOL_NAMES = {tool["name"] for tool in TOOLS}

_RENDERERS = {
    "search_matches": render.render_search_matches,
    "last_match_between": render.render_last_match,
    "head_to_head": render.render_head_to_head,
    "team_stats": render.render_team_stats,
    "team_profile": render.render_team_profile,
    "standings": render.render_standings,
    "biggest_wins": render.render_biggest_wins,
    "competition_stats": render.render_competition_stats,
    "best_records": render.render_best_records,
    "derbies": render.render_derbies,
    "player_search": render.render_player_search,
    "player_club_report": render.render_player_club_report,
    "list_competitions": render.render_list_competitions,
    "list_teams": render.render_list_teams,
}

_DISPATCH = {
    "search_matches": queries.search_matches,
    "last_match_between": queries.last_match_between,
    "head_to_head": queries.head_to_head,
    "team_stats": queries.team_stats,
    "team_profile": queries.team_profile,
    "standings": queries.standings,
    "biggest_wins": queries.biggest_wins,
    "competition_stats": queries.competition_stats,
    "best_records": queries.best_records,
    "derbies": queries.derbies,
    "player_search": queries.player_search,
    "player_club_report": queries.player_club_report,
    "list_competitions": queries.list_competitions,
    "list_teams": queries.list_teams,
}


def tool_summaries() -> str:
    """One line per tool (for README and CLI help)."""
    return "\n".join(
        f"- {t['name']}: {t['description'].splitlines()[0]}" for t in TOOLS
    )


def call_tool(ds: Dataset, name: str, arguments: dict | None) -> dict:
    """
    Execute a tool call.  Returns an MCP tool result:
    ``{"content": [{"type": "text", "text": ...}], "structuredContent": ...}``
    plus ``"isError": True`` when the query itself failed (unknown team,
    bad competition, ...).  Unknown tool names or arguments raise
    :class:`ToolError` (a protocol-level error, per the MCP spec).
    """
    if name not in TOOL_NAMES:
        raise ToolError(
            f"Unknown tool: {name}. Available tools: {', '.join(sorted(TOOL_NAMES))}"
        )
    query = _DISPATCH[name]
    arguments = dict(arguments or {})
    try:
        result = query(ds, **arguments)
    except TypeError as exc:
        # Bad keyword argument for this tool's signature.
        message = str(exc)
        if "keyword argument" in message or "positional argument" in message:
            raise ToolError(f"Invalid arguments for tool '{name}': {message}") from exc
        raise
    if not isinstance(result, dict) or "ok" not in result:
        raise ToolError(f"Tool '{name}' returned an unexpected result")
    text = _RENDERERS[name](result)
    payload = {"content": [{"type": "text", "text": text}], "structuredContent": result}
    if not result.get("ok"):
        payload["isError"] = True
    return payload
